### PR TITLE
Switching Phase Schema to Piecewise Constant. 

### PR DESCRIPTION
**Pull request by weinbe58**
Hardware support for Rabi phase is piecewise constant so this is reflected now in the BloqadeSchema and conversion functions.

### CODE REVIEW

File: lib/BloqadeExpr/test/interface.jl [hunk 1]
```diff
@@ -60,3 +60,18 @@ end
...
+
+
+@testset "get_rydberg_params" begin
+    atoms = [(i,i) for i in 1:10]
+
+    values = [nothing,1,t->t^2,rand(10),[t->rand()*t for i in 1:10]]
+
+    for ϕ in values, Ω in values, Δ in values
+        h = rydberg_h(atoms,ϕ=ϕ,Ω=Ω,Δ=Δ)
+        # catches the this weird edge case
```
Comment (johnzl-777) [inline lib/BloqadeExpr/test/interface.jl:72]:
```suggestion
        # catches weird edge case 
```

File: lib/BloqadeExpr/src/types.jl [hunk 2]
```diff
@@ -536,6 +536,10 @@ YaoAPI.nqubits(::ThreeLevelRydbergConstGates) = 1
 @inline YaoAPI.nqudits(h::RydbergHamiltonian) = nqudits(h.rydberg_term)
 
 
+function Base.:(==)(lhs::DivByTwo{F1},rhs::DivByTwo{F2}) where {F1,F2}
+    return lhs.f == rhs.f
```
Comment (Roger-luo) [inline lib/BloqadeExpr/src/types.jl:540]: I don't think this will always work tho, this only works if the function is identical which is not the case most of the time.

Comment (Roger-luo) [reply lib/BloqadeExpr/src/types.jl:540]: cna you elaborate the use case of having this comparison?

File: lib/BloqadeSchema/src/execute.jl [hunk 3]
```diff
@@ -164,27 +166,18 @@ end
...
-    @debug "Hardware transform report: after linear interpolation ∫dt |Ω(t)-Ω_hw(t)| = $(info.Ω_error) rad"
-    @debug "Hardware transform report: after linear interpolation ∫dt |Δ(t)-Δ_hw(t)| = $(info.Δ_error) rad"
-    @debug "Hardware transform report: mean deviation after rounding positions $(info.mse_atoms) μm"
+    atoms,ϕ,Ω,Δ,δ,Δi = schema_parse_rydberg_fields(h)
 
     violations = validate_analog_params(atoms,ϕ,Ω,Δ,δ,Δi,params.device_capabilities)
 
     if length(violations) > 0
-        throw(ValidationException())
+        show(violations)
```
Comment (Roger-luo) [inline lib/BloqadeSchema/src/execute.jl:174]: this should be cleaned up?

File: lib/BloqadeSchema/src/execute.jl [hunk 4]
```diff
@@ -164,27 +166,18 @@ end
...
+        throw(ValidationException("Failed to convert to Schema, the Hamiltonian doesn't obey hardware constraints."))
     end
 
     ϕ =(
         clocks=ϕ.f.clocks,
-        values=ϕ.f.values
+        values=collect((ϕ.f.values...,ϕ.f.values[end])) # add extra element at the end
```
Comment (Roger-luo) [inline lib/BloqadeSchema/src/execute.jl:180]:
```suggestion
        values=[ϕ.f.values...,ϕ.f.values[end]] # add extra element at the end 
```

**Review by Roger-luo — COMMENTED** [review]
LGTM in general

File: lib/BloqadeSchema/src/execute.jl (hunk 3 above)
Comment (weinbe58) [reply lib/BloqadeSchema/src/execute.jl:174]: I thought we would throw an error after showing all the violations. That was the behavior we had discussed I thought.

File: lib/BloqadeExpr/src/types.jl (hunk 2 above)
Comment (weinbe58) [reply lib/BloqadeExpr/src/types.jl:540]: yeah we don't need this, I thought was a bug in the `==` operator for RydbergHamiltonian but it was not, I forgot to get rid of this!

File: lib/BloqadeExpr/src/types.jl [hunk 5]
```diff
@@ -536,6 +536,10 @@ YaoAPI.nqubits(::ThreeLevelRydbergConstGates) = 1
...
+end
+
```
Comment (weinbe58) [inline lib/BloqadeExpr/src/types.jl:542]:
```suggestion
```

removing function.

File: lib/BloqadeSchema/src/transform.jl [hunk 6]
```diff
@@ -233,7 +235,40 @@ function clip_waveform(wf::Waveform{BloqadeWaveforms.PiecewiseLinear{T,I},T},nam
     )
 end
 
-function hardware_transform_Ω(Ω,device_capabilities::DeviceCapabilities)
+function clip_waveform(wf::Waveform{BloqadeWaveforms.PiecewiseConstant{T},T},name,min_value::T,max_value::T) where {T<:Real,I}
```
Comment (johnzl-777) [inline lib/BloqadeSchema/src/transform.jl:238]:
```suggestion
function clip_waveform(wf::Waveform{BloqadeWaveforms.PiecewiseConstant{T},T},name,min_value::T,max_value::T) where {T<:Real}
```
The `I` parameter isn't being used, can be removed

File: lib/BloqadeSchema/test/execute.jl [hunk 7]
```diff
@@ -83,12 +84,29 @@ end
...
+@testset "conversion to and from schema" begin
+        # Test if code will run without failing
+        Ω = piecewise_linear(;clocks=Float64[0,1,2,3],values=Float64[0,1,1,0])
+        Δ = piecewise_linear(;clocks=Float64[0,1,2,3],values=Float64[1,1,-1,-1])
+        ϕ = constant(;duration=3,value=0)
+        atoms = [(5.0*i,0.0) for i in 1:4]
+        
+        h = rydberg_h(atoms,Ω=Ω,Δ=Δ,ϕ=ϕ)
+        h,info = hardware_transform(h)
+        hh = BloqadeSchema.from_schema(BloqadeSchema.to_schema(h))
```
Comment (johnzl-777) [inline lib/BloqadeSchema/test/execute.jl:108]:
```suggestion
```
`hh` never used, seems to be redundant

File: lib/BloqadeExpr/src/types.jl (hunk 2 above)
Comment (weinbe58) [reply lib/BloqadeExpr/src/types.jl]: Looks like removing this overload broke my test in BloqadeSchema. 

It's only the test that is comparing two RydbergHamiltonian objects.

File: lib/BloqadeSchema/src/execute.jl (hunk 3 above)
Comment (weinbe58) [reply lib/BloqadeSchema/src/execute.jl:174]: I can add a message in the error but I just want the users to look at the violations

File: lib/BloqadeExpr/src/types.jl (hunk 2 above)
Comment (Roger-luo) [reply lib/BloqadeExpr/src/types.jl]: OK, maybe we will need to have it for now, but I think we should be careful when using this comparison in the future as this may not give what you expect

File: lib/BloqadeSchema/src/execute.jl (hunk 3 above)
Comment (Roger-luo) [reply lib/BloqadeSchema/src/execute.jl:174]: sry, I probably wasn't clear, I was thinking to return a violation report in the validate process. So there won't be error at all in this stage, but a validate report object. This object can be empty, then the next pass will take over, or the next pass will complain if the report is not empty.

**Review by johnzl-777 — COMMENTED** [review]
LGTM besides what @Roger-luo 's caught 😃

File: lib/BloqadeSchema/src/execute.jl (hunk 3 above)
Comment (Roger-luo) [reply lib/BloqadeSchema/src/execute.jl:174]: if we print the violations here, then there is no big difference then putting the list of violations into the exception object. But then this is a bit strange as validate function should be expected as a nothrow function (so things do not terminate in the middle, and it would collect all the violations and return back the corresponding location of violation).

also FYI. to print something, it is always preferred to use explicit IO object instead of `stdout` since this won't work well when the output stream is an editor or notebook.

File: lib/BloqadeExpr/src/types.jl (hunk 2 above)
Comment (weinbe58) [reply lib/BloqadeExpr/src/types.jl]: I can just write another test that doesn't rely on the comparison operator.

File: lib/BloqadeSchema/src/execute.jl (hunk 3 above)
Comment (weinbe58) [reply lib/BloqadeSchema/src/execute.jl:174]: > also FYI. to print something, it is always preferred to use explicit IO object instead of `stdout` since this won't work well when the output stream is an editor or notebook.

good point, I will keep that in mind. 

> sry, I probably wasn't clear, I was thinking to return a violation report in the validate process. So there won't be error at all in this stage, but a validate report object. This object can be empty, then the next pass will take over, or the next pass will complain if the report is not empty.
> if we print the violations here, then there is no big difference then putting the list of violations into the exception object. But then this is a bit strange as validate function should be expected as a nothrow function (so things do not terminate in the middle, and it would collect all the violations and return back the corresponding location of violation).

`validate_analog_params ` is the no throw function it simply lists all the violations as we had discussed. The line pointed to here is inside `to_schema` which is the "next function" which is complains. I agree printing the errors isn't great maybe I can pass the `violations` object back to the user through the `ValidationException` object. If I overload the `show` method for `ValidationException` it will allow me to print out the violations in a clean way?

Comment (weinbe58) [reply lib/BloqadeSchema/src/execute.jl]: @Roger-luo checkout this [commit](https://github.com/QuEraComputing/Bloqade.jl/pull/430/commits/ddf34ef48bae4880d14da09df72e251bb75f3581)

Comment (Roger-luo) [reply lib/BloqadeSchema/src/execute.jl]: yeah I think the solution in that commit is better if it's inside `to_schema` since we will have pretty printing from the error object that works in other type of IO too.
